File: flyweight01.py
class FabricaFlyweight:
    def __init__(self):
        self.flyweights = {}

    def obterflyweight(self, codigo):
        if not self.flyweights.get(codigo):
            self.flyweights[codigo] = ProdutoFlyweight(codigo)
        return self.flyweights[codigo]

class ProdutoFlyweight:
    def __init__(self, codigo):
        self.codigo = codigo

    def exibirDetalhes(self, nome, preco):
        print(f"Produto {nome}, Preço R$ {preco}, Código: {self.codigo}")

class Cliente:
    def __init__(self):
        self.FabricaFlyweight = FabricaFlyweight()
        self.carrinho = []

    def adicionarProduto(self, codigo, nome, preco):
        flyweight = self.FabricaFlyweight.obterflyweight(codigo)
        self.carrinho.append({'flyweight': flyweight, 'nome': nome, 'preco': preco})

    def exibirCarrinho(self):
        print("Itens do Carrinho:")
        for item in self.carrinho:
            item['flyweight'].exibirDetalhes(item['nome'], item['preco'])

cliente = Cliente()
cliente.adicionarProduto("001", "Camiseta", 39.90)
cliente.adicionarProduto("002", "Calça", 19.90)
cliente.adicionarProduto("001", "Camiseta", 39.20)
cliente.exibirCarrinho()

File: flyweight01.js
// Flyweight Factory
class FabricaFlyweight {
  constructor() {
    this.flyweights = {}; // Corrigido para 'flyweights' em vez de 'flyweight'
  }

  obterflyweight(codigo) {
    if (!this.flyweights[codigo]) {
      this.flyweights[codigo] = new ProdutoFlyweight(codigo);
    }
    return this.flyweights[codigo];
  }
}

// Flyweight
class ProdutoFlyweight {
  constructor(codigo) {
    this.codigo = codigo;
  }

  exibirDetalhes(nome, preco) {
    console.log(`Produto ${nome}, Preço R$${preco.toFixed(2)}, Código: ${this.codigo}`);
  }
}

// Cliente
class Cliente {
  constructor() {
    this.FabricaFlyweight = new FabricaFlyweight();
    this.carrinho = [];
  }

  adicionarProduto(codigo, nome, preco) {
    const flyweight = this.FabricaFlyweight.obterflyweight(codigo);
    this.carrinho.push({ flyweight, nome, preco });
  }

  exibirCarrinho() {
    console.log(`Itens do Carrinho:`);
    this.carrinho.forEach(item => {
      item.flyweight.exibirDetalhes(item.nome, item.preco);
    });
  }
}

const cliente = new Cliente();
cliente.adicionarProduto("001", "Camiseta", 39.90);
cliente.adicionarProduto("002", "Calça", 29.90);
cliente.adicionarProduto("001", "Camiseta", 30.90); //Reutilizando o mesmo do primeiro
cliente.exibirCarrinho();
